File: src/native/macos/Sources/Updater.swift
import Cocoa

/// In-app auto-updater (docs/auto-update-spec.md §6-7).
///
/// Follows GitHub's `releases/latest` (published stable releases only),
/// downloads the macOS zip + its minisign signature, verifies against the
/// embedded public key, stages the new .app, and swaps it on user-approved
/// relaunch. Background checks are silent and respect the persisted cadence;
/// manual checks always answer.
final class Updater {
    /// Metadata every fetched release carries, even when it matches the
    /// installed version (so the Updates tab can render "Latest: vX (date)"
    /// and a changelog while up to date).
    struct ReleaseInfo {
        let version: String
        let title: String
        let notes: String
        let pageURL: URL
        let publishedAt: Date?

        /// Bullet lines ("- " / "* " / "• ") from the GitHub release body,
        /// lightly de-markdowned. Empty when the body has no list items.
        var changelogBullets: [String] {
            notes.split(whereSeparator: \.isNewline).compactMap { line in
                let trimmed = line.trimmingCharacters(in: .whitespaces)
                for prefix in ["- ", "* ", "• "] where trimmed.hasPrefix(prefix) {
                    return String(trimmed.dropFirst(prefix.count))
                        .replacingOccurrences(of: "**", with: "")
                        .replacingOccurrences(of: "`", with: "")
                        .trimmingCharacters(in: .whitespaces)
                }
                return nil
            }
        }
    }

    struct Release {
        let info: ReleaseInfo
        let zipURL: URL
        let sigURL: URL

        var version: String { info.version }
        var title: String { info.title }
        var notes: String { info.notes }
        var pageURL: URL { info.pageURL }
    }

    enum Phase {
        case idle
        case available(Release)
        case downloading(Release)
        case ready(Release, stagedApp: URL)
        case error(String)
    }

    static let shared = Updater()
    static let phaseChangedNotification = Notification.Name("MouseflareUpdaterPhaseChanged")

    private(set) var phase: Phase = .idle {
        didSet { NotificationCenter.default.post(name: Updater.phaseChangedNotification, object: nil) }
    }

    private let feedURL: URL
    private var checkTimer: Timer?
    private var scheduledIntervalHours = -1
    private var settingsObserver: NSObjectProtocol?

    /// The newest release the feed reported on the last successful check —
    /// kept even when it matches the installed version, so the Updates tab
    /// can show "Latest: vX (date)" and a changelog while up to date.
    private(set) var latestInfo: ReleaseInfo?

    /// The running build's version. CI stamps stable tags (e.g. "0.1.0");
    /// dev builds carry "0.0.0" and never self-update.
    let currentVersion: String
    var isDevBuild: Bool { currentVersion == "0.0.0" }

    private init() {
        let custom = ProcessInfo.processInfo.environment["MOUSEFLARE_UPDATE_FEED"]
        feedURL = custom.flatMap(URL.init(string:))
            ?? URL(string: "https://api.github.com/repos/OffBy1-tech/Mouse-Flare/releases/latest")!
        currentVersion = Bundle.main.object(forInfoDictionaryKey: "CFBundleShortVersionString") as? String ?? "0.0.0"
    }

    // MARK: Background polling

    func startBackgroundChecks() {
        guard !isDevBuild else { return }
        // First quiet check shortly after launch, then per the cadence setting
        DispatchQueue.main.asyncAfter(deadline: .now() + 30) { [weak self] in
            self?.backgroundCheck()
        }
        scheduleTimer()
        // Re-arm when the user changes the check-frequency setting
        settingsObserver = NotificationCenter.default.addObserver(
            forName: SettingsManager.didChangeNotification, object: nil, queue: .main
        ) { [weak self] _ in
            self?.scheduleTimer()
        }
    }

    /// (Re)creates the polling timer to match the persisted cadence.
    /// Interval 0 means manual checks only.
    private func scheduleTimer() {
        let hours = SettingsManager.shared.settings.updateCheckIntervalHours
        guard hours != scheduledIntervalHours else { return }
        scheduledIntervalHours = hours
        checkTimer?.invalidate()
        checkTimer = nil
        guard hours > 0 else { return }
        let timer = Timer(timeInterval: TimeInterval(hours) * 3600, repeats: true) { [weak self] _ in
            self?.backgroundCheck()
        }
        RunLoop.main.add(timer, forMode: .common)
        checkTimer = timer
    }

    private func backgroundCheck() {
        let cfg = SettingsManager.shared.settings
        guard cfg.autoCheckUpdates, cfg.updateCheckIntervalHours > 0 else { return }
        Task { _ = try? await self.check() } // silent on every outcome
    }

    // MARK: Check

    private struct Asset: Decodable {
        let name: String
        let browser_download_url: String
    }
    private struct Feed: Decodable {
        let tag_name: String
        let name: String?
        let body: String?
        let html_url: String
        let assets: [Asset]
        let published_at: String?
    }

    /// Fetches the feed. Returns the newer release, or nil when up to date.
    /// Updates `phase` but never downgrades an in-flight or staged update.
    @discardableResult
    func check() async throws -> Release? {
        var request = URLRequest(url: feedURL)
        request.setValue("application/vnd.github+json", forHTTPHeaderField: "Accept")
        let (data, response) = try await URLSession.shared.data(for: request)
        guard let http = response as? HTTPURLResponse, http.statusCode == 200 else {
            throw UpdaterError.feedUnavailable
        }

        let feed = try JSONDecoder().decode(Feed.self, from: data)

        let version = feed.tag_name.hasPrefix("v") ? String(feed.tag_name.dropFirst()) : feed.tag_name
        guard let pageURL = URL(string: feed.html_url) else { throw UpdaterError.feedUnavailable }
        let info = ReleaseInfo(
            version: version,
            title: feed.name ?? "Mouseflare v\(version)",
            notes: feed.body ?? "",
            pageURL: pageURL,
            publishedAt: feed.published_at.flatMap { ISO8601DateFormatter().date(from: $0) }
        )
        await MainActor.run {
            self.latestInfo = info
            SettingsManager.shared.settings.lastUpdateCheck = Date()
        }

        guard version != currentVersion, !isDevBuild else {
            await MainActor.run {
                if case .available = self.phase { self.phase = .idle }
                if case .error = self.phase { self.phase = .idle }
            }
            return nil
        }

        guard
            let zip = feed.assets.first(where: { $0.name == "Mouseflare-macOS.zip" }),
            let sig = feed.assets.first(where: { $0.name == "Mouseflare-macOS.zip.minisig" }),
            let zipURL = URL(string: zip.browser_download_url),
            let sigURL = URL(string: sig.browser_download_url)
        else {
            throw UpdaterError.assetsMissing
        }

        let release = Release(info: info, zipURL: zipURL, sigURL: sigURL)
        await MainActor.run {
            switch self.phase {
            case .downloading, .ready:
                break // never downgrade an in-flight or staged update
            default:
                self.phase = .available(release)
            }
        }
        return release
    }

    // MARK: Download, verify & stage

    func downloadAndStage(_ release: Release) async {
        await MainActor.run { self.phase = .downloading(release) }
        do {
            let (zipData, zipResponse) = try await URLSession.shared.data(from: release.zipURL)
            let (sigData, sigResponse) = try await URLSession.shared.data(from: release.sigURL)
            guard (zipResponse as? HTTPURLResponse)?.statusCode == 200,
                  (sigResponse as? HTTPURLResponse)?.statusCode == 200,
                  let sigText = String(data: sigData, encoding: .utf8) else {
                throw UpdaterError.downloadFailed
            }

            // The security boundary: verify before touching the payload
            try Minisign.verify(data: zipData, signatureFile: sigText)

            // Stage on the same volume as the installed app so the final swap
            // is an atomic same-volume move.
            let bundleURL = Bundle.main.bundleURL
            let staging = try FileManager.default.url(
                for: .itemReplacementDirectory,
                in: .userDomainMask,
                appropriateFor: bundleURL,
                create: true
            )
            let zipPath = staging.appendingPathComponent("Mouseflare-macOS.zip")
            try zipData.write(to: zipPath)

            let unzip = Process()
            unzip.executableURL = URL(fileURLWithPath: "/usr/bin/ditto")
            unzip.arguments = ["-x", "-k", zipPath.path, staging.path]
            try unzip.run()
            unzip.waitUntilExit()
            guard unzip.terminationStatus == 0 else { throw UpdaterError.unpackFailed }

            let stagedApp = staging.appendingPathComponent("Mouseflare.app")
            let stagedBinary = stagedApp.appendingPathComponent("Contents/MacOS/Mouseflare")
            guard FileManager.default.fileExists(atPath: stagedBinary.path) else {
                throw UpdaterError.unpackFailed
            }

            await MainActor.run { self.phase = .ready(release, stagedApp: stagedApp) }
        } catch {
            await MainActor.run { self.phase = .error(error.localizedDescription) }
        }
    }

    // MARK: Install & relaunch

    /// Whether the running app can replace itself in place.
    var canSelfInstall: Bool {
        let bundleURL = Bundle.main.bundleURL
        guard bundleURL.pathExtension == "app" else { return false }
        guard !bundleURL.path.contains("/AppTranslocation/") else { return false }
        return FileManager.default.isWritableFile(atPath: bundleURL.deletingLastPathComponent().path)
    }

    /// Swaps the staged app into place and relaunches. Restores the previous
    /// version if the swap fails partway.
    func installAndRelaunch(stagedApp: URL) throws {
        let fm = FileManager.default
        let bundleURL = Bundle.main.bundleURL
        guard canSelfInstall else { throw UpdaterError.cannotSelfInstall }

        let backupDir = try fm.url(
            for: .itemReplacementDirectory,
            in: .userDomainMask,
            appropriateFor: bundleURL,
            create: true
        )
        let backup = backupDir.appendingPathComponent("Mouseflare-previous.app")

        try fm.moveItem(at: bundleURL, to: backup)
        do {
            try fm.moveItem(at: stagedApp, to: bundleURL)
        } catch {
            try? fm.moveItem(at: backup, to: bundleURL) // roll back
            throw error
        }
        try? fm.removeItem(at: backup)

        // Relaunch the new version after this process exits
        let relaunch = Process()
        relaunch.executableURL = URL(fileURLWithPath: "/bin/sh")
        relaunch.arguments = ["-c", "sleep 0.5; /usr/bin/open -n \"$0\"", bundleURL.path]
        try relaunch.run()

        DispatchQueue.main.async {
            NSApp.terminate(nil)
        }
    }

    enum UpdaterError: Error, LocalizedError {
        case feedUnavailable
        case assetsMissing
        case downloadFailed
        case unpackFailed
        case cannotSelfInstall

        var errorDescription: String? {
            switch self {
            case .feedUnavailable: return "The release feed could not be reached."
            case .assetsMissing: return "The latest release has no macOS artifacts."
            case .downloadFailed: return "The update could not be downloaded."
            case .unpackFailed: return "The downloaded update could not be unpacked."
            case .cannotSelfInstall: return "Mouseflare cannot update itself from this location — download the update manually."
            }
        }
    }
}

File: src/native/windows/Core/Updater.cs
using System;
using System.IO;
using System.IO.Compression;
using System.Linq;
using System.Net.Http;
using System.Reflection;
using System.Text.Json;
using System.Threading;
using System.Threading.Tasks;

namespace Mouseflare.Core
{
    /// <summary>
    /// In-app auto-updater (docs/auto-update-spec.md §6, §8), mirroring the
    /// macOS Updater.swift: follows GitHub's releases/latest (published stable
    /// only), downloads zip + minisign signature, verifies against the embedded
    /// key, stages, and swaps via the rename-a-running-exe dance on restart.
    ///
    /// Deliberately WPF-free so the check/verify/stage pipeline is testable on
    /// any platform; only InstallAndRestart is Windows-specific.
    /// </summary>
    public sealed class Updater
    {
        public enum UpdatePhase { Idle, Available, Downloading, Ready, Error }

        public sealed class ReleaseInfo
        {
            public string Version = "";
            public string Title = "";
            public string Notes = "";
            public string ZipUrl = "";
            public string SigUrl = "";
            public string PageUrl = "";
            public DateTime? PublishedAt;  // GitHub published_at
            public string[] Changelog = Array.Empty<string>(); // bullet lines from the release body
        }

        public static readonly Updater Shared = new();

        public UpdatePhase Phase { get; private set; } = UpdatePhase.Idle;
        public ReleaseInfo? AvailableRelease { get; private set; }
        public string? StagedDirectory { get; private set; }
        public string? ErrorMessage { get; private set; }

        /// <summary>Raised (on a thread-pool thread) whenever Phase changes.</summary>
        public event Action? PhaseChanged;

        /// <summary>Gate for background checks; wire to the persisted setting.</summary>
        public Func<bool> AutoCheckEnabled = () => true;

        /// <summary>Background check cadence in hours (0 = manual only); wire to the persisted setting.</summary>
        public Func<int> CheckIntervalHours = () => 6;

        /// <summary>UTC time of the last completed feed fetch (survives restarts via settings.json).</summary>
        public DateTime? LastCheckedUtc { get; private set; }

        /// <summary>Newest stable release seen on the feed — populated even when up to date.</summary>
        public ReleaseInfo? LatestSeenRelease { get; private set; }

        /// <summary>Restores the persisted last-check time at startup (never overwrites a live one).</summary>
        public void SeedLastChecked(DateTime? utc)
        {
            if (LastCheckedUtc == null) LastCheckedUtc = utc;
        }

        public readonly string CurrentVersion;
        public bool IsDevBuild => CurrentVersion == "0.0.0";

        private readonly Uri _feedUrl;
        private readonly HttpClient _http;
        private Timer? _checkTimer;

        private Updater()
        {
            var custom = Environment.GetEnvironmentVariable("MOUSEFLARE_UPDATE_FEED");
            _feedUrl = Uri.TryCreate(custom, UriKind.Absolute, out var uri)
                ? uri
                : new Uri("https://api.github.com/repos/OffBy1-tech/Mouse-Flare/releases/latest");

            _http = new HttpClient();
            _http.DefaultRequestHeaders.UserAgent.ParseAdd("Mouseflare-Updater"); // required by the GitHub API
            _http.DefaultRequestHeaders.Accept.ParseAdd("application/vnd.github+json");

            var informational = Assembly.GetExecutingAssembly()
                .GetCustomAttribute<AssemblyInformationalVersionAttribute>()?.InformationalVersion ?? "0.0.0";
            CurrentVersion = informational.Split('+')[0];
        }

        // ---- Background polling ----

        public void StartBackgroundChecks()
        {
            if (IsDevBuild) return;
            // A fast tick with an interval gate, so cadence changes apply
            // without rescheduling the timer. 0 hours = manual checks only.
            _checkTimer = new Timer(
                _ =>
                {
                    int hours = CheckIntervalHours();
                    if (hours <= 0 || !AutoCheckEnabled()) return;
                    if (LastCheckedUtc is { } last && DateTime.UtcNow - last < TimeSpan.FromHours(hours)) return;
                    _ = CheckSilently();
                },
                null,
                TimeSpan.FromSeconds(30),
                TimeSpan.FromMinutes(30)
            );
        }

        private async Task CheckSilently()
        {
            try { await CheckAsync(); } catch { /* background checks never nag */ }
        }

        // ---- Check ----

        /// <summary>
        /// Fetches the feed. Returns the differing release, or null when up to
        /// date. Always records LastCheckedUtc and LatestSeenRelease; updates
        /// Phase but never downgrades an in-flight/staged update.
        /// </summary>
        public async Task<ReleaseInfo?> CheckAsync()
        {
            ReleaseInfo release;
            using (var doc = JsonDocument.Parse(await _http.GetStringAsync(_feedUrl)))
            {
                release = ParseRelease(doc.RootElement);
            }

            LatestSeenRelease = release;
            LastCheckedUtc = DateTime.UtcNow;

            if (IsDevBuild || release.Version == CurrentVersion)
            {
                if (Phase is UpdatePhase.Available or UpdatePhase.Error)
                {
                    SetPhase(UpdatePhase.Idle, release: null);
                }
                return null;
            }

            if (release.ZipUrl.Length == 0 || release.SigUrl.Length == 0)
            {
                throw new InvalidOperationException("The latest release has no Windows artifacts.");
            }

            if (Phase is not (UpdatePhase.Downloading or UpdatePhase.Ready))
            {
                SetPhase(UpdatePhase.Available, release);
            }
            return release;
        }

        private static ReleaseInfo ParseRelease(JsonElement root)
        {
            string tag = root.GetProperty("tag_name").GetString() ?? "";
            string version = tag.StartsWith("v") ? tag.Substring(1) : tag;

            string? zipUrl = null, sigUrl = null;
            if (root.TryGetProperty("assets", out var assets))
            {
                foreach (var asset in assets.EnumerateArray())
                {
                    string name = asset.GetProperty("name").GetString() ?? "";
                    string url = asset.GetProperty("browser_download_url").GetString() ?? "";
                    if (name == "Mouseflare-Windows.zip") zipUrl = url;
                    if (name == "Mouseflare-Windows.zip.minisig") sigUrl = url;
                }
            }

            string notes = root.TryGetProperty("body", out var b) ? b.GetString() ?? "" : "";
            DateTime? published = root.TryGetProperty("published_at", out var pub)
                && pub.ValueKind == JsonValueKind.String
                && DateTime.TryParse(pub.GetString(), null, System.Globalization.DateTimeStyles.AdjustToUniversal, out var when)
                ? when : null;

            return new ReleaseInfo
            {
                Version = version,
                Title = root.TryGetProperty("name", out var n) ? n.GetString() ?? $"Mouseflare v{version}" : $"Mouseflare v{version}",
                Notes = notes,
                ZipUrl = zipUrl ?? "",
                SigUrl = sigUrl ?? "",
                PageUrl = root.GetProperty("html_url").GetString() ?? "",
                PublishedAt = published,
                Changelog = ParseChangelog(notes),
            };
        }

        /// <summary>Bullet lines ("- ", "* ", "• ") from the release body, markers stripped.</summary>
        private static string[] ParseChangelog(string body) =>
            body.Split('\n')
                .Select(line => line.Trim())
                .Where(line => line.StartsWith("- ") || line.StartsWith("* ") || line.StartsWith("• "))
                .Select(line => line.Substring(2).Trim().Trim('*').Trim())
                .Where(line => line.Length > 0)
                .Take(8)
                .ToArray();

        // ---- Download, verify & stage ----

        public async Task DownloadAndStageAsync(ReleaseInfo release)
        {
            SetPhase(UpdatePhase.Downloading, release);
            try
            {
                byte[] zip = await _http.GetByteArrayAsync(release.ZipUrl);
                string sig = await _http.GetStringAsync(release.SigUrl);

                // The security boundary: verify before unpacking anything
                MinisignVerifier.Verify(zip, sig);

                string staging = Path.Combine(InstallDirectory, "update-staging");
                if (Directory.Exists(staging)) Directory.Delete(staging, recursive: true);
                Directory.CreateDirectory(staging);

                using (var archive = new ZipArchive(new MemoryStream(zip)))
                {
                    archive.ExtractToDirectory(staging);
                }
                if (!File.Exists(Path.Combine(staging, "Mouseflare.exe")))
                {
                    throw new InvalidOperationException("The downloaded update is missing Mouseflare.exe.");
                }

                StagedDirectory = staging;
                SetPhase(UpdatePhase.Ready, release);
            }
            catch (Exception ex)
            {
                ErrorMessage = ex.Message;
                SetPhase(UpdatePhase.Error, release: null);
            }
        }

        // ---- Install (Windows rename dance) ----

        public static string InstallDirectory => AppContext.BaseDirectory;

        public bool CanSelfInstall
        {
            get
            {
                try
                {
                    string probe = Path.Combine(InstallDirectory, ".mf-write-probe");
                    File.WriteAllText(probe, "");
                    File.Delete(probe);
                    return true;
                }
                catch { return false; }
            }
        }

        /// <summary>
        /// Swaps the staged files into place, starts the new exe, and returns
        /// so the caller can shut this instance down. Never overwrites a file
        /// in place: every existing destination is renamed to .old first — a
        /// rename succeeds even for the running exe and (usually) for files the
        /// runtime has memory-mapped, where an overwrite would throw "in use by
        /// another process" (seen in the field with a lazily-loaded .pdb).
        /// Non-essential files that still cannot be replaced are skipped;
        /// anything else rolls the whole swap back.
        /// </summary>
        public void InstallAndRestart(string stagedDirectory)
        {
            string exePath = Environment.ProcessPath
                ?? throw new InvalidOperationException("Cannot locate the running executable.");
            var renamed = new System.Collections.Generic.List<(string Original, string Old)>();

            static bool IsNonEssential(string path) =>
                Path.GetExtension(path).Equals(".pdb", StringComparison.OrdinalIgnoreCase);

            try
            {
                foreach (string source in Directory.EnumerateFiles(stagedDirectory, "*", SearchOption.AllDirectories))
                {
                    string relative = Path.GetRelativePath(stagedDirectory, source);
                    string destination = Path.Combine(InstallDirectory, relative);
                    Directory.CreateDirectory(Path.GetDirectoryName(destination)!);
                    try
                    {
                        if (File.Exists(destination))
                        {
                            string old = destination + ".old";
                            if (File.Exists(old)) File.Delete(old);
                            File.Move(destination, old);
                            renamed.Add((destination, old));
                        }
                        File.Copy(source, destination);
                    }
                    catch when (IsNonEssential(destination))
                    {
                        // e.g. a .pdb the runtime holds without delete-sharing:
                        // stale debug symbols beat a failed update
                    }
                }
            }
            catch
            {
                // Roll back so the current install keeps working
                foreach (var (original, old) in renamed)
                {
                    try { if (File.Exists(original)) File.Delete(original); } catch { }
                    try { if (!File.Exists(original)) File.Move(old, original); } catch { }
                }
                throw;
            }
            try { Directory.Delete(stagedDirectory, recursive: true); } catch { }

            System.Diagnostics.Process.Start(new System.Diagnostics.ProcessStartInfo
            {
                FileName = exePath,
                UseShellExecute = true,
                WorkingDirectory = InstallDirectory,
            });
        }

        /// <summary>Startup housekeeping: remove leftovers from the previous swap.</summary>
        public static void CleanupAfterUpdate()
        {
            try
            {
                foreach (string old in Directory.EnumerateFiles(InstallDirectory, "*.old", SearchOption.AllDirectories))
                {
                    try { File.Delete(old); } catch { /* may still be releasing; next launch gets it */ }
                }
                string staging = Path.Combine(InstallDirectory, "update-staging");
                if (Directory.Exists(staging)) Directory.Delete(staging, recursive: true);
            }
            catch { }
        }

        private void SetPhase(UpdatePhase phase, ReleaseInfo? release)
        {
            Phase = phase;
            AvailableRelease = release ?? (phase == UpdatePhase.Idle ? null : AvailableRelease);
            if (phase != UpdatePhase.Error) ErrorMessage = null;
            if (phase != UpdatePhase.Ready) StagedDirectory = phase == UpdatePhase.Downloading ? StagedDirectory : null;
            PhaseChanged?.Invoke();
        }
    }
}
